File: MVC/mvc.py
import sqlite3
 
class DefectModel(object):
 
    def getDefectList(self, component):
        query = '''select ID from defects where Component = '%s' ''' %component
        defectlist = self._dbselect(query)
        list = []
        for row in defectlist:
            list.append(row[0])
        return list
 
    def getSummary(self, id):
        query = '''select summary from defects where ID = '%d' ''' % id
        summary = self._dbselect(query)
        for row in summary:
            return row[0]
 
    def _dbselect(self, query):
        connection = sqlite3.connect('./TMS')
        cursorObj = connection.cursor()
        cursorObj.execute(query)
        return cursorObj.fetchall()
 
class DefectView(object):
 
    def summary(self, summary, defectid):
        print "#### Defect Summary for defect# %d####\n%s" % (defectid,summary)

    def defectList(self, list, category):
        print "#### Defect List for %s ####" % category
        for defect in list:
            print defect

class FileDefectView(object):
 
    def __init__(self, filename):
        self.fh = open(filename, 'wb')
    def __del__(self):
        self.fh.close()

    def summary(self, summary, defectid):
        self.fh.write( "#### Defect Summary for defect# %d####\n%s" % (defectid,summary))

    def defectList(self, list, category):
        self.fh.write( "\n#### Defect List for %s ####" % category)
        for defect in list:
            self.fh.write('\n' + str(defect))
 
class Controller(object):
 
    def __init__(self, model=DefectModel(), view=DefectView()):
        self.model = model
        self.view  = view
 
    def getDefectSummary(self, defectid):
        summary_data = self.model.getSummary(defectid)
        self.view.summary(summary_data, defectid)
 
    def getDefectList(self, component):
        defectlist_data = self.model.getDefectList(component)
        self.view.defectList(defectlist_data, component)

File: MVC/mvc.rb
#!/usr/bin/env ruby

# TODO:
#
# Introduce Mocking and unit test for this MVC stuff
# Relevant Model in MVC? ping. own thread? for 
# slack, IRC, restAPI sinatra??
#
#
class DefectModel
  def initialize
    puts '0'
  end

  def get_summary(defect_id)
    puts '1'
  end

  def get_defect_list(component)
    puts '2'
  end

end

class DefectView

  def initialize
    puts '3'
  end

  def summary(summary_data, defectid)
    puts '4'
  end

  def defect_list(defeclist_data, component)
    puts '5'
  end

end

class Controller

  def initialize( model=nil, view=nil)
    @model ||= DefectModel.new
    @view  ||=  DefectView.new 
  end

  def get_defect_summary(defect_id)
    summary_data = @model.get_summary(defect_id)
    @view.summary(summary_data, defect_id)
  end

  def get_defect_list(component)
    defeclist_data = @model.get_defect_list(component)
    @view.defect_list(defeclist_data, component)
  end

end


if __FILE__ == $PROGRAM_NAME

  begin

  controller = Controller.new
  controller.get_defect_summary(34)

  controller.get_defect_list(42)
 
  rescue => msg
    puts msg
    puts msg.backtrace

  end
end